Unit 2.3 LangGraph : Release 11/03/25 (#271)
huggingface/agents-course: units/en/unit2/langgraph/first_graph.mdx

# Building Your First LangGraph

Now that we understand the building blocks, let's put them into practice by building our first functional graph. We'll implement Alfred's email processing system, where he needs to:

1. Read incoming emails
2. Classify them as spam or legitimate
3. Draft a preliminary response for legitimate emails

> 4. Send information to Mr Wayne when legitimate (printing only)

Reviewer: ```suggestion
4. Send information to Mr. Wayne when legitimate (printing only)
```

> This example demonstrates how to structure a workflow with LangGraph that involves LLM based decision making. WHile this can't be considered an Agent as no tool is involved, this section focuses more on learning the langGraph framework than Agents.

Reviewer: ```suggestion
This example demonstrates how to structure a workflow with LangGraph that involves LLM-based decision-making. While this can't be considered an Agent as no tool is involved, this section focuses more on learning the LangGraph framework than Agents.
```

> You can follow the process [in this notebook](https://huggingface.co/agents-course/notebooks/resolve/main/unit2/langgraph/mail_sorting.ipynb) or **run the code yourself**.

Reviewer: ```suggestion
<Tip>
You can follow the code in <a href="https://huggingface.co/agents-course/notebooks/resolve/main/unit2/langgraph/mail_sorting.ipynb" target="_blank">this notebook</a> that you can run using Google Colab.
</Tip>
```

Reviewer: Following examples from previous units

}

# Process the legitimate email
print("\nProcessing legitimate email...")
legitimate_result = compiled_graph.invoke({
    "email": legitimate_email,
    "is_spam": None,
    "spam_reason": None,
    "email_category": None,
    "draft_response": None,
    "messages": []
})

# Process the spam email
print("\nProcessing spam email...")
spam_result = compiled_graph.invoke({
    "email": spam_email,
    "is_spam": None,
    "spam_reason": None,
    "email_category": None,
    "draft_response": None,
    "messages": []
})
```

## Step 6: Inspecting Our Mail Sorting Agent with Langfuse 📡

As Alfred fine-tunes the Main Sorting Agent, he's growing weary of debugging its runs. Agents, by nature, are unpredictable and difficult to inspect. But since he aims to build the ultimate Spam Detection Agent and deploy it in production, he needs robust traceability for future monitoring and analysis. 

To do this, Alfred can use an observability tool such as [Langfuse](https://langfuse.com/) to trace and monitor the agent.

First, we pip install Langfuse:  
```python

> %pip install -q langfuse

Reviewer: Here, the pip instruction starts with `%`, while the pip instruction above does not start the same way.
